plonky3/src/circuit_builder.rs

```diff
@@ -0,0 +1,193 @@
+
+        let res = match e {
+            AlgebraicExpression::Reference(r) => {
+                let poly_id = r.poly_id;
+
+                let row = match r.next {
+                    true => matrix.row_slice(1),
+                    false => matrix.row_slice(0),
+                };
+
+                // witness columns indexes are unchanged, fixed ones are offset by `commitment_count`
+                let index = match poly_id.ptype {
+                    PolynomialType::Committed => {
+                        assert!(
+                            r.poly_id.id < self.commitment_count as u64,
+                            "Plonky3 expects `poly_id` to be contiguous"
+                        );
+                        r.poly_id.id as usize
+                    }
+                    PolynomialType::Constant => {
+                        assert!(
+                            r.poly_id.id < self.constant_count as u64,
+                            "Plonky3 expects `poly_id` to be contiguous"
+                        );
+                        self.commitment_count + r.poly_id.id as usize
+                    }
+                    PolynomialType::Intermediate => {
+                        unreachable!("intermediate polynomials should have been inlined")
+                    }
+                };
+
+                row[index].into()
+            }
+            AlgebraicExpression::PublicReference(_) => unimplemented!(
+                "public references are not supported inside algebraic expressions in plonky3"
+            ),
+            AlgebraicExpression::Number(n) => AB::Expr::from(cast_to_goldilocks(*n)),
+            AlgebraicExpression::BinaryOperation(left, op, right) => {
+                let left = self.to_plonky3_expr(left, builder);
+                let right = self.to_plonky3_expr(right, builder);
+
+                match op {
+                    AlgebraicBinaryOperator::Add => left + right,
+                    AlgebraicBinaryOperator::Sub => left - right,
+                    AlgebraicBinaryOperator::Mul => left * right,
+                    AlgebraicBinaryOperator::Pow => {
+                        unreachable!("exponentiations should have been evaluated")
+                    }
+                }
+            }
+            AlgebraicExpression::UnaryOperation(op, e) => {
+                let e: <AB as AirBuilder>::Expr = self.to_plonky3_expr(e, builder);
+
+                match op {
+                    AlgebraicUnaryOperator::Minus => -e,
+                }
+            }
+            AlgebraicExpression::Challenge(challenge) => {
+                unimplemented!("Challenge API for {challenge:?} not accessible in plonky3")
+            }
+        };
+        res
+    }
+}
+
+impl<'a, T: FieldElement> BaseAir<Val> for PowdrCircuit<'a, T> {
+    fn width(&self) -> usize {
+        self.commitment_count + self.constant_count
+    }
+
+    fn preprocessed_trace(&self) -> Option<RowMajorMatrix<Val>> {
```

Comment:
I don't know why this is actually in the `BaseAir` trait (doesn't seem to be called by Plonky3?), but from the name I think they mean this to return fixed columns only.

In SP1, they actually don't implement this function (they panic when it's called) and instead implement [this trait](https://github.com/succinctlabs/sp1/blob/main/core/src/air/machine.rs), which has both `generate_preprocessed_trace()` and `generate_trace()`, which return fixed and committed columns.